libraries/chain/fork_database.cpp

```diff
@@ -298,16 +299,15 @@ namespace eosio::chain {
    }
 
    template<class BSP>
-   void fork_database_t<BSP>::reset_root( const bhs_t& root_bhs ) {
+   void fork_database_t<BSP>::reset_root( const bsp_t& root_bsp ) {
       std::lock_guard g( my->mtx );
-      my->reset_root_impl(root_bhs);
+      my->reset_root_impl(root_bsp);
    }
 
    template<class BSP>
-   void fork_database_impl<BSP>::reset_root_impl( const bhs_t& root_bhs ) {
+   void fork_database_impl<BSP>::reset_root_impl( const bsp_t& root_bsp ) {
       index.clear();
-      root = std::make_shared<bs_t>();
-      static_cast<bhs_t&>(*root) = root_bhs;
+      root = root_bsp;
       bs_accessor_t::set_valid(*root, true);
```

Comment:
Should only set valid if this was a version 1 `root_bsp` from `open_impl`. Otherwise, it should already be `valid`. Could assert it is valid and set the valid in `open_impl` if a `block_header_state` was read instead of a `block_state`.

Reply:
I'm not quite sure why, but it appears that `validated` is not set on root after loading a fork_db when running tests with this PR's code, se the `set_valid` is necessary.

Reply:
Set valid is only necessary when loading a version 1 fork database which does not have all the `block_state` (doesn't read in `validated`.). I think you should move the set valid into `open_impl` and assert in `reset_root_impl`.

Reply:
> Set valid is only necessary when loading a version 1 fork database which does not have all the `block_state` (doesn't read in `validated`.). I think you should move the set valid into `open_impl` and assert in `reset_root_impl`.

I tried that but many tests fail when I do that.

Reply:
Glancing at the tests, I think I know why. I can address that in #2141

Reply:
I'm curious to know what the issue could be?

Reply:
Mock block_state rely on this to set valid. Those need to set valid outside.

Reply:
> Mock block_state rely on this to set valid. Those need to set valid outside.

Thanks, that makes sense.